File: wallendar-app/app/src/test/java/com/td/wallendar/home/ui/HomePresenterTest.java
package com.td.wallendar.home.ui;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import static io.reactivex.Single.error;
import static io.reactivex.Single.just;

import com.td.wallendar.TestSchedulerProvider;
import com.td.wallendar.models.ApplicationUser;
import com.td.wallendar.models.Debt;
import com.td.wallendar.models.Group;
import com.td.wallendar.repositories.interfaces.ApplicationUsersRepository;
import com.td.wallendar.repositories.interfaces.DebtsRepository;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

public class HomePresenterTest {

    private HomeView homeView;
    private GroupsRepository groupsRepository;
    private DebtsRepository debtsRepository;
    private ApplicationUsersRepository applicationUsersRepository;
    private SchedulerProvider schedulerProvider = new TestSchedulerProvider();

    private HomePresenter homePresenter;

    private final Group fakeGroup = new Group(1L, "group title", new ApplicationUser(), new HashSet<>(), new HashSet<>(), new HashSet<>(), new HashSet<>());
    private List<Group> groups = new ArrayList<>();

    private ApplicationUser tdallas = new ApplicationUser(1L, "tdallas@itba.edu.ar", "Tomas", "Dallas", "");
    private ApplicationUser tdorado = new ApplicationUser(1L, "tdorado@itba.edu.ar", "Tomas", "Dorado", "");

    private final Debt fakeDebt = new Debt(
            1L,
            tdallas,
            tdorado,
            25d,
            1L
    );
    private List<Debt> debts = new ArrayList<>();


    @Before
    public void setup() {
        groups.add(fakeGroup);
        debts.add(fakeDebt);

        homeView = mock(HomeView.class);
        groupsRepository = mock(GroupsRepository.class);
        debtsRepository = mock(DebtsRepository.class);
        applicationUsersRepository = mock(ApplicationUsersRepository.class);

        homePresenter = new HomePresenter(homeView, groupsRepository, debtsRepository, applicationUsersRepository, schedulerProvider);
    }

    @Test
    public void whenViewIsAttachedThenGroupsShouldBeFetchedAndLoad_Ok() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(groups));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(Collections.emptyList()));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView).bindGroups(groups);
    }

    @Test
    public void whenViewIsAttachedThenDebtsForUserShouldBeFetchedAndLoad_Ok() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(Collections.emptyList()));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(debts));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView).bindDebts(debts);
    }

    @Test
    public void whenViewIsAttachedThenUserShouldBeFetchedAndLoad_Ok() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(Collections.emptyList()));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(Collections.emptyList()));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView).bindProfile(tdallas);
    }

    @Test
    public void whenViewIsAttachedThenGroupsAndUserAndDebtsShouldBeFetchedAndLoaded_Ok() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(groups));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(debts));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView).bindProfile(tdallas);
        verify(homeView).bindGroups(groups);
        verify(homeView).bindDebts(debts);
    }

    // Errors
    @Test
    public void whenViewIsAttachedThenGroupsShouldBeFetchedAndLoad_WithErrors() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(error(new Exception("Fake Exception")));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(debts));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView, times(1)).errorGettingGroups();

        verify(homeView, times(1)).bindDebts(debts);
        verify(homeView, times(1)).bindProfile(tdallas);
    }

    @Test
    public void whenViewIsAttachedThenDebtsForUserShouldBeFetchedAndLoad_WithErrors() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(groups));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(error(new Exception("Fake Exception")));
        when(applicationUsersRepository.getUser(1L)).thenReturn(just(tdallas));

        homePresenter.onViewAttached(1L);

        verify(homeView).errorGettingBalances();
        verify(homeView, times(1)).bindProfile(tdallas);
        verify(homeView, times(1)).bindGroups(groups);
    }

    @Test
    public void whenViewIsAttachedThenUserShouldBeFetchedAndLoad_WithErrors() {
        when(groupsRepository.getGroupsByUser(1L)).thenReturn(just(Collections.emptyList()));
        when(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(just(Collections.emptyList()));
        when(applicationUsersRepository.getUser(1L)).thenReturn(error(new Exception("Fake exception")));

        homePresenter.onViewAttached(1L);

        verify(homeView).errorGettingUser();
        verify(homeView, times(1)).bindGroups(Collections.emptyList());
        verify(homeView, times(1)).bindDebts(Collections.emptyList());
    }


}

File: wallendar-app/app/src/test/java/com/td/wallendar/home/ui/HomePresenterTest.kt
package com.td.wallendar.home.ui

import com.td.wallendar.TestSchedulerProvider
import com.td.wallendar.models.Group
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import java.lang.Exception
import java.util.ArrayList
import java.util.HashSet

class HomePresenterTest {
    private var homeView: HomeView? = null
    private var groupsRepository: GroupsRepository? = null
    private var debtsRepository: DebtsRepository? = null
    private var applicationUsersRepository: ApplicationUsersRepository? = null
    private val schedulerProvider: SchedulerProvider? = TestSchedulerProvider()
    private var homePresenter: HomePresenter? = null
    private val fakeGroup: Group? = Group(1L, "group title", ApplicationUser(), HashSet(), HashSet(), HashSet(), HashSet())
    private val groups: MutableList<Group?>? = ArrayList()
    private val tdallas: ApplicationUser? = ApplicationUser(1L, "tdallas@itba.edu.ar", "Tomas", "Dallas", "")
    private val tdorado: ApplicationUser? = ApplicationUser(1L, "tdorado@itba.edu.ar", "Tomas", "Dorado", "")
    private val fakeDebt: Debt? = Debt(
            1L,
            tdallas,
            tdorado,
            25.0,
            1L
    )
    private val debts: MutableList<Debt?>? = ArrayList<Debt?>()
    @Before
    fun setup() {
        groups.add(fakeGroup)
        debts.add(fakeDebt)
        homeView = Mockito.mock<HomeView?>(HomeView::class.java)
        groupsRepository = Mockito.mock<GroupsRepository?>(GroupsRepository::class.java)
        debtsRepository = Mockito.mock<DebtsRepository?>(DebtsRepository::class.java)
        applicationUsersRepository = Mockito.mock<ApplicationUsersRepository?>(ApplicationUsersRepository::class.java)
        homePresenter = HomePresenter(homeView, groupsRepository, debtsRepository, applicationUsersRepository, schedulerProvider)
    }

    @Test
    fun whenViewIsAttachedThenGroupsShouldBeFetchedAndLoad_Ok() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just(groups))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(emptyList()))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).bindGroups(groups)
    }

    @Test
    fun whenViewIsAttachedThenDebtsForUserShouldBeFetchedAndLoad_Ok() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just<MutableList<Group?>?>(emptyList()))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(debts))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).bindDebts(debts)
    }

    @Test
    fun whenViewIsAttachedThenUserShouldBeFetchedAndLoad_Ok() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just<MutableList<Group?>?>(emptyList()))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(emptyList()))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).bindProfile(tdallas)
    }

    @Test
    fun whenViewIsAttachedThenGroupsAndUserAndDebtsShouldBeFetchedAndLoaded_Ok() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just(groups))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(debts))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).bindProfile(tdallas)
        Mockito.verify<HomeView?>(homeView).bindGroups(groups)
        Mockito.verify<HomeView?>(homeView).bindDebts(debts)
    }

    // Errors
    @Test
    fun whenViewIsAttachedThenGroupsShouldBeFetchedAndLoad_WithErrors() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.error<MutableList<Group?>?>(Exception("Fake Exception")))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(debts))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).errorGettingGroups()
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindDebts(debts)
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindProfile(tdallas)
    }

    @Test
    fun whenViewIsAttachedThenDebtsForUserShouldBeFetchedAndLoad_WithErrors() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just(groups))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.error<MutableList<Debt?>?>(Exception("Fake Exception")))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.just<ApplicationUser?>(tdallas))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).errorGettingBalances()
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindProfile(tdallas)
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindGroups(groups)
    }

    @Test
    fun whenViewIsAttachedThenUserShouldBeFetchedAndLoad_WithErrors() {
        Mockito.`when`<Single<MutableList<Group?>?>?>(groupsRepository.getGroupsByUser(1L)).thenReturn(Single.just<MutableList<Group?>?>(emptyList()))
        Mockito.`when`<Single<MutableList<Debt?>?>?>(debtsRepository.getTotalDebtsByUserId(1L)).thenReturn(Single.just<MutableList<Debt?>?>(emptyList()))
        Mockito.`when`<Single<ApplicationUser?>?>(applicationUsersRepository.getUser(1L)).thenReturn(Single.error<ApplicationUser?>(Exception("Fake exception")))
        homePresenter.onViewAttached(1L)
        Mockito.verify<HomeView?>(homeView).errorGettingUser()
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindGroups(emptyList())
        Mockito.verify<HomeView?>(homeView, Mockito.times(1)).bindDebts(emptyList())
    }
}